pm/queue.go

```diff
@@ -0,0 +1,158 @@
+// If a sender's max float is insufficient to cover the face value of a ticket it is added to the queue.
+// A ticket is pulled from the queue by the recipient when a sender has sufficient max float to cover
+// the next ticket in the queue
+//
+// Based off of: https://github.com/lightningnetwork/lnd/blob/master/htlcswitch/queue.go
+type ticketQueue struct {
+	queue []*SignedTicket
+
+	// queueLen is an internal length counter that keeps track
+	// of the size of the queue. We maintain this counter instead
+	// of reading len(queue) in order to avoid acquiring the main lock
+	// used by the queue loop goroutine
+	queueLen int32
+
+	// cond is a conditional variable that is used by the main
+	// queue loop goroutine to wait for new tickets added to the queue
+	cond *sync.Cond
+
+	// availableMaxFloat is a channel that the queue uses
+	// to receive the latest available max float value
+	availableMaxFloat chan *big.Int
+
+	// redeemable is a channel that a recipient will receive
+	// redeemable tickets on as a sender's max float becomes
+	// sufficient to cover the face value of tickets
+	redeemable chan *SignedTicket
+
+	quit chan struct{}
+}
+
+func newTicketQueue() *ticketQueue {
+	return &ticketQueue{
+		cond:              sync.NewCond(&sync.Mutex{}),
+		availableMaxFloat: make(chan *big.Int),
+		redeemable:        make(chan *SignedTicket),
+		quit:              make(chan struct{}),
+	}
+}
+
+// Start initiates the main queue loop goroutine for processing tickets
+func (q *ticketQueue) Start() {
+	go q.startQueueLoop()
+}
+
+// Stop signals the ticketQueue to gracefully shutdown
+func (q *ticketQueue) Stop() {
+	close(q.quit)
+}
+
+// Add adds a ticket to the queue
+func (q *ticketQueue) Add(ticket *SignedTicket) {
+	// Lock conditional variable while adding to the queue
+	q.cond.L.Lock()
+	q.queue = append(q.queue, ticket)
+	atomic.AddInt32(&q.queueLen, 1)
+	q.cond.L.Unlock()
+
+	// Signal that there are tickets in the queue
+	q.cond.Signal()
+}
+
+// SignalMaxFloat signals to the queue the latest max float for the sender
+// An external caller should call this method whenever a ticket redemption confirms
+// with the most up-to-date max float for the sender
+func (q *ticketQueue) SignalMaxFloat(amount *big.Int) {
+	// If the queue is empty do not signal a new max float
+	// value because there are no tickets to process
+	if q.Length() == 0 {
+		return
+	}
+
+	select {
+	case q.availableMaxFloat <- amount:
+	case <-q.quit:
+		return
+	}
+}
+
+// Redeemable returns a channel that a consumer can use to receive tickets that
+// should be redeemed
+func (q *ticketQueue) Redeemable() chan *SignedTicket {
+	return q.redeemable
+}
+
+// Length returns the current length of the queue
+func (q *ticketQueue) Length() int32 {
+	return atomic.LoadInt32(&q.queueLen)
+}
+
+func (q *ticketQueue) startQueueLoop() {
+	for {
+		// Lock and wait until the queue is non-empty
+		q.cond.L.Lock()
+		for len(q.queue) == 0 {
+			q.cond.Wait()
```

Comment:
Are you suggesting to use a channel as the queue itself instead of using the condvar + a slice? Whenever a max float update is signalled, we peek at the head of the queue and if the ticket at the head of the queue cannot be covered by the max float, we leave the queue as-is. I guess if we wanted to use a channel we could try to replicate this behavior by reading the ticket from the channel and if it cannot be covered by the max float we pass the ticket back into the channel. The one difference is that at the moment the queue remains unchanged when a max float is insufficient to cover the head of the queue while the described channel approach would pop the head of the queue and add it to the tail in the same scenario.

Reply:
@j0sh 

I ended up taking a different approach in f46945f than what is described in the above comment. I swapped the condvar for an unbuffered channel which the `(*queue).Add()` sends tickets into. The queue loop uses a select statement to receive on the unbuffered channel and if there is nothing to receive it will proceed to the rest of the loop. Let me know what you think!

Reply:
That looks closer to what I was thinking of. Will add some notes inline.

Reply:
BTW if it seems more straightforward to maintain the old approach with condvars then by all means stay with that.

(Offside rant, but this is one of the things I dislike about golang's concurrency constructs: the language doesn't give you a lot of tools to safely use those constructs. The experience of reasoning through their use is still a very raw process and not a huge step up from traditional primitives for non-trivial cases. It doesn't help that most of the behaviors are 'design choices' so they have to be learned and aren't as amenable to deduction from first principles.)

Reply:
Yeah I think this is a case where after looking at a few other concurrency safe queues for inspiration the use of condvar felt a bit more intuitive than trying to use channels. Though I guess trying to figure out how to cleanly replace condvar here with channels is a useful thought exercise...